sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pyx.tp

```diff
@@ -158,6 +165,47 @@ cdef class DatasetsPair{{name_suffix}}:
         # TODO: add "with gil: raise" here when supporting Cython 3.0
         return -1
 
+
+@final
+cdef class PrecomputedDistanceMatrix{{name_suffix}}(DatasetsPair{{name_suffix}}):
+    """A subclass of DatasetsPair 
+    
+    Parameters: must receive precomputed_distance: ndarray of shape
+        (n_samples_X, n_samples_Y),
+        Must be C-contiguous. 
+    """
+
+    def __init__(
+        self,
+        const {{INPUT_DTYPE_t}}[:, ::1] precomputed_distance,  
+    ):
+        super().__init__(
+            distance_metric=DistanceMetric{{name_suffix}}(),
+            n_features=0,  
+        )
+        # This array has already been checked.
+        self.precomputed_distance = precomputed_distance
```

Comment:
This explains why you needed to take the GIL.
```suggestion
        self.distance_matrix = precomputed_distance
```

Reply:
Thanks.